### DISCUSSION TITLE
Clarification, RE #523 ("Not working with .NET Core 3 on Alpine Linux")

### DISCUSSION THREAD

**Discussion by jnickg** [dlemstra/Magick.NET#735] [Help] [answered]
I read #523 and #556, but am getting the same issue. It is not trivial for me to switch from the alpine container, because it causes failures/crashes elsewhere in my app that I have not yet investigated.

The reason I am filing a new issue for this it to ask a clarifying question: is it possible this might somehow be a naming issue?

Here's my exception from my Heroku logs:
```
2020-09-17T06:44:41.126413+00:00 app[web.1]:       An unhandled exception has occurred while executing the request.
2020-09-17T06:44:41.126503+00:00 app[web.1]: System.TypeInitializationException: The type initializer for 'NativeMagickImageCollection' threw an exception.
2020-09-17T06:44:41.126526+00:00 app[web.1]:  ---> System.DllNotFoundException: Unable to load shared library 'Magick.Native-Q8-x64.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: Error loading shared library libMagick.Native-Q8-x64.dll: No such file or directory
```

I tried searching for the file(s) called out in the exception (`libMagick.Native-Q8-x64.dll` and `Magick.Native-Q8-x64.dll`) on my system, by adding `RUN find / -iname *magick*` in my Dockerfile at the end of the build. Here is what I found:
```
/app/Magick.Native-Q8-x64.dll.so
/app/Magick.NET.Core.dll
/app/Magick.NET-Q8-AnyCPU.dll
```
I see `Magick.Native-Q8-x64.dll.so`, but no `libMagick.Native-Q8-x64.dll` or `Magick.Native-Q8-x64.dll`. Sorry if this is something obvious that I'm missing (maybe there is a fuzzy name lookup, hard-coded exception messages, an automatic rename, or something else). Still, I  thought it was worth asking, just to check.

-----

**Note:** Similarly, immediately after NuGet restore, before I've copied artifacts here's what I see on the system:
```
Step 11/22 : RUN find / -iname *magick*
 ---> Running in 933d24fef7b4
/root/.nuget/packages/magick.net-q8-anycpu
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/magick.net-q8-anycpu.7.21.1.nupkg.sha512
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/Magick.NET.icon.png
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/lib/net20/Magick.NET-Q8-AnyCPU.dll
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/lib/netstandard20/Magick.NET-Q8-AnyCPU.dll
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/lib/netstandard13/Magick.NET-Q8-AnyCPU.dll
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/lib/net40/Magick.NET-Q8-AnyCPU.dll
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/magick.net-q8-anycpu.nuspec
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/magick.net-q8-anycpu.7.21.1.nupkg
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/runtimes/win-x86/native/Magick.Native-Q8-x86.dll
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/runtimes/osx-x64/native/Magick.Native-Q8-x64.dll.dylib
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/runtimes/linux-x64/native/Magick.Native-Q8-x64.dll.so
/root/.nuget/packages/magick.net-q8-anycpu/7.21.1/runtimes/win-x64/native/Magick.Native-Q8-x64.dll
/root/.nuget/packages/magick.net.core
/root/.nuget/packages/magick.net.core/4.1.0/Magick.NET.icon.png
/root/.nuget/packages/magick.net.core/4.1.0/magick.net.core.nuspec
/root/.nuget/packages/magick.net.core/4.1.0/lib/net20/Magick.NET.Core.dll
/root/.nuget/packages/magick.net.core/4.1.0/lib/netstandard20/Magick.NET.Core.dll
/root/.nuget/packages/magick.net.core/4.1.0/lib/netstandard13/Magick.NET.Core.dll
/root/.nuget/packages/magick.net.core/4.1.0/lib/net40/Magick.NET.Core.dll
/root/.nuget/packages/magick.net.core/4.1.0/magick.net.core.4.1.0.nupkg
/root/.nuget/packages/magick.net.core/4.1.0/magick.net.core.4.1.0.nupkg.sha512
/root/.local/share/NuGet/v3-cache/1ca707a4d90792ce8e42453d4e350886a0fdaa4d$ps:_api.nuget.org_v3_index.json/nupkg_magick.net.core.4.1.0.dat
/root/.local/share/NuGet/v3-cache/1ca707a4d90792ce8e42453d4e350886a0fdaa4d$ps:_api.nuget.org_v3_index.json/nupkg_magick.net-q8-anycpu.7.21.1.dat
/root/.local/share/NuGet/v3-cache/1ca707a4d90792ce8e42453d4e350886a0fdaa4d$ps:_api.nuget.org_v3_index.json/list_magick.net.core.dat
/root/.local/share/NuGet/v3-cache/1ca707a4d90792ce8e42453d4e350886a0fdaa4d$ps:_api.nuget.org_v3_index.json/list_magick.net-q8-anycpu.dat
Removing intermediate container 933d24fef7b4
```

**Comment by jnickg** [comment]
Somebody else suggested this might be the case, [here](https://github.com/dlemstra/Magick.NET/issues/523#issuecomment-539157849)

**Comment by dlemstra** [comment, marked as answer]
I did some more research and it turns out that I will need to create a linux-musl specific binary. Don't know when I will have time for that though.

**Reply by jnickg** [reply]
Thanks for clarifying! I will figure out whatever issues I am having switching over to the container that's known to work.

**Reply by dlemstra** [reply]
I made some progress this weekend to create a `musl` build and this should become available in the next release. The binary has already been created and the last step is to integrate this into this project. Will try to do that tonight.

**Reply by jnickg** [reply]
You rock!!!

**Reply by dlemstra** [reply]
You can download a beta here: https://github.com/dlemstra/Magick.NET/actions from the latest build. You could download the packages file unzip it and put one of the packages in your local NuGet feed. Would you be willing to test this for me?

**Reply by jnickg** [reply]
Sorry, did not check my GH notifications for a long time. I did get my app working on the non-`musl` version, so I no longer need this fix personally. That said, if you still need help testing this I can do so -- I should have a commit with the `musl` version of my Dockerfile.
